clang/lib/3C/CastPlacement.cpp

```diff
@@ -18,107 +18,244 @@
+  auto CastStrs = getCastString(SrcInt, SrcExt, DstInt, DstExt);
+
   // If E is already a cast expression, we will try to rewrite the cast instead
   // of adding a new expression.
   if (auto *CE = dyn_cast<CStyleCastExpr>(E->IgnoreParens())) {
     SourceRange CastTypeRange(CE->getLParenLoc(), CE->getRParenLoc());
-    Writer.ReplaceText(CastTypeRange, CastPrefix);
+    Writer.ReplaceText(CastTypeRange, CastStrs.first.substr(1));
   } else {
+    // First try to insert the cast prefix and suffix around the extression in
+    // the source code.
     bool FrontRewritable = Writer.isRewritable(E->getBeginLoc());
     bool EndRewritable = Writer.isRewritable(E->getEndLoc());
     if (FrontRewritable && EndRewritable) {
-      bool FFail = Writer.InsertTextAfterToken(E->getEndLoc(), ")");
-      bool EFail = Writer.InsertTextBefore(E->getBeginLoc(), "(" + CastPrefix);
+      bool BFail = Writer.InsertTextBefore(E->getBeginLoc(), CastStrs.first);
+      bool EFail = Writer.InsertTextAfterToken(E->getEndLoc(), CastStrs.second);
       assert("Locations were rewritable, fail should not be possible." &&
-             !FFail && !EFail);
+             !BFail && !EFail);
     } else {
-      // This means we failed to insert the text at the end of the RHS.
-      // This can happen because of Macro expansion.
-      // We will see if this is a single expression statement?
-      // If yes, then we will use parent statement to add ")"
+      // Sometimes we can't insert the cast around the expression due to macros
+      // getting in the way. In these cases, we can sometimes replace the entire
+      // expression source with a new string containing the orginal expression
+      // and the cast.
       auto CRA = CharSourceRange::getTokenRange(E->getSourceRange());
       auto NewCRA = clang::Lexer::makeFileCharRange(
           CRA, Context->getSourceManager(), Context->getLangOpts());
       std::string SrcText = clang::tooling::getText(CRA, *Context);
-      // Only insert if there is anything to write.
+      // This doesn't always work either. We can't rewrite if the cast needs to
+      // be placed fully inside a macro rather than around a macro or on an
+      // argument to the macro.
       if (!SrcText.empty())
-        Writer.ReplaceText(NewCRA, "(" + CastPrefix + SrcText + ")");
+        Writer.ReplaceText(NewCRA, CastStrs.first + SrcText + CastStrs.second);
     }
   }
 }
+
+bool CastLocatorVisitor::exprHasCast(Expr *E) const {
+  return isa<CastExpr>(E) || ExprsWithCast.find(E) != ExprsWithCast.end();
+}
+
+bool CastLocatorVisitor::VisitCastExpr(CastExpr *C) {
+  if (!isa<ImplicitCastExpr>(C)) {
+    Expr *Sub = ignoreCheckedCImplicit(C->getSubExpr());
+    ExprsWithCast.insert(Sub);
+  }
+  return true;
+}
+
+Expr *CastLocatorVisitor::ignoreCheckedCImplicit(Expr *E) {
+  Expr *Old = nullptr;
+  Expr *New = E;
+  while (Old != New) {
+    Old = New;
+    New = Old->IgnoreExprTmp()->IgnoreImplicit();
+  }
+  return New;
+}
```

Comment:
This can go in `Utils.cpp` ? Also, the two above it seems suspiciously general-purpose too, but for the set.